CONTRIBUTING.md

```diff
@@ -1,76 +1,401 @@
+
+**Many pack-files:** You can split a fresh clone into multiple pack-files of size at most
+16MB by running `git repack -adfF --max-pack-size=16m`. See the
+[`git repack` documentation](https://git-scm.com/docs/git-repack) for more information.
+You can count the number of pack-files using `ls .git/objects/pack/*.pack | wc -l`.
+
+**Many loose objects:** If you already split your repository into multiple pack-files, then
+you can pick one to split into loose objects using `cat .git/objects/pack/[id].pack | git unpack-objects`;
+delete the `[id].pack` and `[id].idx` files after this. You can count the number of loose
+bjects using `ls .git/objects/??/* | wc -l`.
+
+**Deep history:** Usually large repositories also have deep histories, but you can use the
+[test-many-commits-1m repo](https://github.com/cirosantilli/test-many-commits-1m/) to
+target deep histories without the overhead of many objects. One issue with this repository:
+there are no merge commits, so you will need to use a different repository to test a "wide"
+commit history.
+
+**Large Index:** You can generate a large index and repo by using the scripts in
+`t/perf/repos`.  There are two scripts. `many-files.sh` which will generate a repo with
+same tree and blobs but different paths.  Using `many-files.sh -d 5 -w 10 -f 9` will create
+a repo with ~1 million entries in the index. `inflate-repo.sh` will use an existing repo
+and copy the current work tree until it is a specified size.
+
+Test Your Changes on Linux
+--------------------------
+
+It can be important to work directly on the [core Git codebase](https://github.com/git/git),
+such as a recent commit into the `master` or `next` branch that has not been incorporated
+into Git for Windows. Also, it can help to run functional and performance tests on your
+code in Linux before submitting patches to the Linux-focused mailing list.
+
+### Using the Windows Subsystem for Linux
+
+The [Windows Subsystem for Linux (WSL)](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
+allows you to [install Ubuntu Linux as an app](https://www.microsoft.com/en-us/store/p/ubuntu/9nblggh4msv6)
+that can run Linux executables on top of the Windows kernel. Internally,
+Linux syscalls are interpreted by the WSL, everything else is plain Ubuntu.
+
+First, open WSL (either type "Bash" in Cortana, or execute "bash.exe" in a CMD window).
+Then install the prerequisites, and `git` for the initial clone:
+
+```
+sudo apt-get update
+sudo apt-get install git gcc make libssl-dev libcurl4-openssl-dev libexpat-dev tcl tk gettext git-email
+```
+
+Then, clone and build:
+
+```
+git clone https://github.com/git-for-windows/git
+cd git
+git remote add -f upstream https://github.com/git/git
+make
+```
+
+Be sure to clone into `/home/[user]/` and not into any folder under `/mnt/?/` or your build
+will fail due to colons in file names.
+
+### Using a Linux Virtual Machine with Hyper-V

```

Comment:
Maybe also mention that there are alternatives to this MS product? Virtualbox (Oracle, Free, Open source) and VMWare (Free product) are the two big ones.

Reply:
Careful with suggesting those alternatives. We want this guide to be practical, and suggesting to use Virtualbox (which I personally am unable to get to run because it needs Hyper-V to be *disabled*, while I need it *enabled* for Docker) may very well lead to a lot of totally unnecessary frustration.

And honestly, if you want alternatives to be mentioned, why not open a PR to derrickstolee:contributing? That would put the onus on you, of course, but if you want it...?

Reply:
I'm all for practicality, but MS Hyper-V requires Windows Pro, which maybe not everyone wants to splurge on.

Reply:
As I said, if you want alternatives to be mentioned, you can make that happen by opening a Pull Request into @derrickstolee's branch.

I agree with you, @avivey, that it would be good to mention alternatives, in particular for users who cannot use Hyper-V. I also agree with @derrickstolee that Hyper-V is a very convenient way to get a Linux VM on Windows (if you can use Hyper-V).

And since @derrickstolee seems to be familiar with Hyper-V, and since you seem to be familiar with other options, it looks as if you were the ideal person to document at least one other option.

Reply:
I can do a writeup about "running linux vm in windows"; I don't think it should go in this document. In fact, I don't think it should be written down explicitly, because these things change all the time, and there's plenty of good instructions online already - just saying "google linux vm in windows" should be enough, considering the skill level required for that vs "changing code in git" (and even "format patch for email").
If it were me, I'd just say:
> run your tests in linux. Here are the common options:
> 1. Ubuntu For Windows: a bit slow, but works fine other than that.
> 2. Hyper-V: Excellent choice if you have Windows 10 Pro.
> 3. VirtualBox: Great open-source solution, a little under-polished.
> 3. VMWare Player: Great free commercial tool.
> 5. A friend that "knows linux": best options for just running a few tests.
>
> Pick your vm engine and install:
> - Download the ISO.....

Reply:
I don't like this. This basically tells contributors "you know what? Just go and figure out all of this yourself, we don't care that you have to struggle with this."

But I do care.

I don't want Windows-based developers who are unfamiliar with Linux to struggle with this step, as testing your changes on Linux can point out tons of issues that would otherwise be pointed out by humans on the Git mailing list, with always feels quite a bit like personal criticism, and can be avoided. These issues include (but are not limited to): test scripts not being marked as executable on Linux, line ending issues, missing `#ifdef` guards for code that only compiles cleanly on Windows, etc.

Reply:
The idea of this document is to give a path for new contributors. It should not mention _every_ path, but it should give at least one. Users can figure out which parts they know how to do and which parts they need guidance to complete.

I will reword to mention that other VM applications exist, and that Hyper-V may not be installed with your version of Windows. I will keep my explicit instructions for Hyper-V, though.